api/src/main/java/com/github/retrooper/packetevents/protocol/player/storage/FastUserStorage.java

```diff
@@ -0,0 +1,112 @@
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+
+/**
+ * Fast User Storage - An API usage class
+ * This class provides efficient runtime value association with the User class (the association exists for as along as the User doesn't quit - it's non-permanent).
+ * It can be used for many different purposes, such as, but not limited to:
+ * 1. Storing the plugin's packet handler for the User even before UUID is known. This could be achieved for example via:
+ *
+ * <pre> {@code
+ *  private final TypedStorageValueId<MyUser> identifier = TypedStorageValueId.identifierFor(this.myPluginInstance);
+ *
+ *  @Override
+ *  public void onUserConnect(UserConnectEvent event) {
+ *      event.getUser().storeRuntime(this.identifier, new MyUser());//associate
+ *  }
+ *
+ *  @Override
+ *  public void onPacketReceive(PacketReceiveEvent event) {
+ *      MyUser user = event.getUser().getStored(this.identifier);
+ *      user.packetHandler().handle(event);
+ *  }
+ * }</pre>
+ * <p>
+ * 2. Replacing a Map - Even tho Maps are extremely fast, their performance can slightly decrease with size. This Storage class offers constant O(1)
+ * performance, independently to the amount of Users or the length of the data array. With 100 users, the performance of this class was often found to be
+ * about ~35% faster than that of a UUID Map.
+ *
+ * @author ShadowOfHeaven
+ */
+
+public final class FastUserStorage {
+
+    //Ensure the array's element thread visibility by using MethodHandles
+    //It was found to be slightly faster than VarHandles with excessive amount of calls
+    private static final MethodHandle setElement = MethodHandles.arrayElementSetter(Object[].class);//array, index, value
+    private static final MethodHandle getElement = MethodHandles.arrayElementGetter(Object[].class);//array, index
+    private static final Map<Object, Integer> pluginToId = new ConcurrentHashMap<>(2);
+    private static volatile int idCounter;
+    private volatile Object[] data;
+
+    public FastUserStorage() {
+        this.data = new Object[idCounter];
+    }
+
+    public void store(StorageIdBase identifier, Object value) {
+        this.store0(identifier.id(), value);
+    }
+
+    private void store0(int index, Object value) {
+        if (index >= data.length) this.regrow();
+        try {
+            setElement.invokeExact(this.data, index, value);//must be invoked with an int
+        } catch (Throwable e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public Object get(StorageValueId identifier) {
+        return this.get0(identifier.id());
+    }
+
+    public <T> T get(TypedStorageValueId<T> identifier) {
+        return (T) this.get0(identifier.id());
+    }
+
+    private Object get0(int index) {
+        if (index >= data.length) {
+            this.regrow();
```

Comment:
why are you growing here?

Reply:
On any data access, be it write or read, if the class that creates the identifier is initialized later on the initial data size won't be enough. The grow happens only when we're implicitly told that this has happened. It's also safe to do this, since the id is provided internally only